src/shogun/lib/DynamicArray.h

```diff
@@ -13,30 +13,872 @@
+		{
+			m_array = orig.m_array;
+			num_elements = orig.num_elements;
+			dim1_size = orig.dim1_size;
+			dim2_size = orig.dim2_size;
+			dim3_size = orig.dim3_size;
+
+			return *this;
+		}
+
+		/** shuffles the array (not thread safe!) */
+		inline void shuffle()
+		{
+			for (index_t i = 0; i <= num_elements - 1; ++i)
+				CMath::swap(
+				    m_array[i], m_array[CMath::random(i, num_elements - 1)]);
+		}
+
+		/** shuffles the array with external random state */
+		inline void shuffle(CRandom* rand)
+		{
+			for (index_t i = 0; i <= num_elements - 1; ++i)
+				CMath::swap(
+				    m_array[i], m_array[rand->random(i, num_elements - 1)]);
+		}
+
+		/** display this array */
+		inline void display_array()
+		{
+			if (get_name())
+				SG_PRINT(
+				    "DynamicArray '%s' of size: %dx%dx%d\n", get_name(),
+				    dim1_size, dim2_size, dim3_size)
+			else
+				SG_PRINT(
+				    "DynamicArray of size: %dx%dx%d\n", dim1_size, dim2_size,
+				    dim3_size)
+
+			for (int32_t k = 0; k < dim3_size; k++)
+				for (int32_t i = 0; i < dim1_size; i++)
+				{
+					SG_PRINT("element(%d,:,%d) = [ ", i, k)
+					for (int32_t j = 0; j < dim2_size; j++)
+						SG_PRINT("%1.1f,", (float32_t)element(i, j, k))
+					SG_PRINT(" ]\n")
+				}
+		}
+
+		/** display array's size */
+		inline void display_size()
+		{
+			SG_PRINT(
+			    "DynamicArray of size: %dx%dx%d\n", dim1_size, dim2_size,
+			    dim3_size)
+		}
+
+		/** @return object name */
+		virtual const char* get_name() const
+		{
+			return "DynamicArray";
+		}
+
+		/** Can (optionally) be overridden to pre-initialize some member
+		 *  variables which are not PARAMETER::ADD'ed.  Make sure that at
+		 *  first the overridden method BASE_CLASS::LOAD_SERIALIZABLE_PRE
+		 *  is called.
+		 *
+		 *  @exception ShogunException Will be thrown if an error
+		 *                             occurres.
+		 */
+		virtual void load_serializable_pre() throw(ShogunException)
+		{
+			CSGObject::load_serializable_pre();
+
+			resize_array(get_num_elements());
+		}
+
+		/** Can (optionally) be overridden to pre-initialize some member
+		 *  variables which are not PARAMETER::ADD'ed.  Make sure that at
+		 *  first the overridden method BASE_CLASS::SAVE_SERIALIZABLE_PRE
+		 *  is called.
+		 *
+		 *  @exception ShogunException Will be thrown if an error
+		 *                             occurres.
+		 */
+		virtual void save_serializable_pre() throw(ShogunException)
+		{
+			CSGObject::save_serializable_pre();
+			resize_array(get_num_elements());
+		}
+
+		virtual CSGObject* clone()
+		{
+			CDynamicArray* cloned = (CDynamicArray*)CSGObject::clone();
+			// Since the array vector is registered with
+			// current_num_elements as size (see parameter
+			// registration) the cloned version has less memory
+			// allocated than known to dynarray. We fix this here.
+			// cloned->num_elements = cloned->array_size;
+			return cloned;
+		}
+
+	private:
+		/** register parameters */
+		virtual void init()
+		{
```

Comment:
@vigsterkr Hi, could you tell me something about set_generic<>()?
I got bunch of error like
```

/Users/mikeling/shogun/build/src/shogun/base/class_list.cpp:815:110: error: use
      of class template 'CDynamicArray' requires template arguments
  ...g) { return g == PT_NOT_GENERIC? new CDynamicArray(): NULL; }
                                          ^
/Users/mikeling/shogun/src/shogun/base/DynArray.h:34:33: note: template is
      declared here
        template<class U> friend class CDynamicArray;
        ~~~~~~~~~~~~~~~~~              ^
/Users/mikeling/shogun/build/src/shogun/base/class_list.cpp:815:131: error: 
      expected ':'
  ...g) { return g == PT_NOT_GENERIC? new CDynamicArray(): NULL; }
                                                               ^
                                                               : 
/Users/mikeling/shogun/build/src/shogun/base/class_list.cpp:815:104: note: to
      match this '?'
  ...g) { return g == PT_NOT_GENERIC? new CDynamicArray(): NULL; }
                                    ^
/Users/mikeling/shogun/build/src/shogun/base/class_list.cpp:815:131: error: 
      expected expression
  ...g) { return g == PT_NOT_GENERIC? new CDynamicArray(): NULL; }
```